### PR TITLE
Add Iterator to Tree

### PR DESCRIPTION
**Pull request by arielshaqed**
Iterates over all values in a tree, putting a "new part" header as an explicit iterated value.
Makes it easier to write Apply (and hopefully Diff) easier.

### CODE REVIEW

**Comment by arielshaqed** [comment]
Please have a look whether you think this is The Right Way :tm: (or at least A Right Way) to write an abstraction to help us write Apply and Diff.

File: graveler/committed/tree/apply.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			if ok := source.Next(); !ok {
+				sourceDone = true
+			}
+		}
+	}
+	for !diffsDone {
+		diffValue := diffs.Value()
+		if err := writer.WriteRecord(*diffValue); err != nil {
+			return fmt.Errorf("write added record: %w", err)
+		}
```
Comment (guy-har) [inline graveler/committed/tree/apply.go:85]: Maybe we should check for tombstone here  (even though it shouldn't happen)

File: graveler/committed/tree/tree.go [hunk 2]
```diff
@@ -19,6 +19,22 @@ type Tree struct {
 	Parts []Part
 }
 
+// PartsAndValuesIterator iterates over all part headers and values of a tree, allowing
+// seeking by entire parts.
+type PartsAndValuesIterator interface {
+	// Next moves to iterator to look at the next value in the current part, or a header
+	// for the next part if the current part is over.
```
Comment (guy-har) [inline graveler/committed/tree/tree.go:26]: moves the Iterator

File: graveler/committed/tree/parts_and_values.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		return false
+	}
+	pvi.parts = pvi.parts[1:]
+	pvi.it = nil
+	return true
+}
+
+func (pvi *partsAndValuesIterator) Next() bool {
+	if pvi.it != nil {
+		ok := pvi.it.Next()
```
Comment (johnnyaug) [inline graveler/committed/tree/parts_and_values.go:27]: I like `if pvi.it.Next()` - but matter of personal taste, I guess

**Review by guy-har — COMMENTED** [review]
Looks Good, I like the new iterator

**Review by johnnyaug — COMMENTED** [review]
Looks very good

File: graveler/committed/tree/parts_and_values.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
+package tree
+
+import "github.com/treeverse/lakefs/graveler"
+
+type partsAndValuesIterator struct {
```
Comment (nopcoder) [inline graveler/committed/tree/parts_and_values.go:5]: Suggestion shorter name - TreeValueIterator

File: graveler/committed/tree/parts_and_values.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	repo  Repo
+	parts []Part
+	it    graveler.ValueIterator
+	err   error
+}
```
Comment (nopcoder) [inline graveler/committed/tree/parts_and_values.go:10]: This iterator also behaves a bit differently as the expected behavior from our current Iterators / Scanners are to call Next, on true you can access Value to get the current iterator value.

**Review by nopcoder — APPROVED** [review]
love it, think we need to align our iterators/scanners in the Graveler project and we will be fine 👍

**Review by arielshaqed — COMMENTED** [review]
Removing code for `Apply` to pull just the iterator.  `Apply` will come in the next PR.  Meanwhile, this is no longer a draft, so please request changes if you don't want this iterator.

File: graveler/committed/tree/parts_and_values.go (hunk 4 above)
Comment (johnnyaug) [reply graveler/committed/tree/parts_and_values.go:5]: I suggest: TreeIterator for this one,
and the other iterator (which only returns values) will be TreeValueIterator

**Review by johnnyaug — APPROVED** [review]

**Review by guy-har — APPROVED** [review]

File: graveler/committed/tree/parts_and_values.go (hunk 4 above)
Comment (johnnyaug) [reply graveler/committed/tree/parts_and_values.go:5]: And since we're under the tree -pun not intended- package, how about just Iterator.

**Comment by arielshaqed** [comment]
Thanks for reviews!

System test failed because of some DNS lookup failure flakiness (attn @ozkatz ...), so merging by admin force -- this code is not yet on any path exercised by the system test!

**Pull request merged**